client/jetpack-connect/controller.js

```diff
@@ -141,57 +154,72 @@ export function persistMobileAppFlow( context, next ) {
 }
 
 export function setMasterbar( context, next ) {
+	debug( 'controller: setMasterbar', context.params );
 	if ( config.isEnabled( 'jetpack/connect/mobile-app-flow' ) ) {
 		const masterbarToggle = retrieveMobileRedirect() ? hideMasterbar() : showMasterbar();
 		context.store.dispatch( masterbarToggle );
 	}
 	next();
 }
 
-export function connect( context, next ) {
-	const { path, pathname, params, query } = context;
-	const { type = false, interval } = params;
-	const analyticsPageTitle = get( type, analyticsPageTitleByType, 'Jetpack Connect' );
-	const planSlug = getPlanSlugFromFlowType( type, interval );
-
-	// Not clearing the plan here, because other flows can set the cookie before arriving here.
-	planSlug && storePlan( planSlug );
-	recordPageView( pathname, analyticsPageTitle );
+export function loginBeforeJetpackSearch( context, next ) {
+	debug( 'controller: loginBeforeJetpackSearch', context.params );
+	const { params, path } = context;
+	const { type } = params;
+	const isLoggedOut = ! getCurrentUserId( context.store.getState() );
 
-	context.primary = (
-		<JetpackConnect
-			ctaFrom={ query.cta_from /* origin tracking params */ }
-			ctaId={ query.cta_id /* origin tracking params */ }
-			locale={ params.locale }
-			path={ path }
-			type={ type }
-			url={ query.url }
-			forceRemoteInstall={ query.forceInstall }
-		/>
-	);
+	// Log in to WP.com happens at the start of the flow for Search products
+	// ( to facilitate site selection ).
+	if ( JETPACK_SEARCH_PRODUCTS.includes( type ) && isLoggedOut ) {
+		return page( login( { isNative: true, isJetpack: true, redirectTo: path } ) );
+	}
 	next();
 }
 
-// Purchase Jetpack Search
-export function purchase( context, next ) {
+export function connect( context, next ) {
```

Comment:
I merged the `purchase` and `connect` controllers because they were almost the same. They only differed in the rendered component: `JetpackConnect` or `SearchPurchase`.